File: leetcode/src/main/java/leetcode/string/ValidAnagram.java
package leetcode.string;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.Map;
import java.util.Queue;
import java.util.HashMap;
import java.util.PriorityQueue;

/**
 * https://leetcode.com/problems/valid-anagram/
 *
 * Nếu hai chuỗi có các phần tử bằng nhau thì bằng nhau (không tính thứ tự)
 */
public class ValidAnagram {

    public boolean isAnagram(String s1, String s2) {
        if (s1.length() != s2.length()) {
            return false;
        }
        Map<Character, Integer> charCount = new HashMap<>();
        for (char c : s1.toCharArray()) {
            charCount.put(c, charCount.getOrDefault(c, 0) + 1);
        }
        for (char c : s2.toCharArray()) {
            // Ký tự không xuất hiện trong chuỗi s1 hoặc đã trừ hết lần xuất hiện
            if (!charCount.containsKey(c) || charCount.get(c) == 0) {
                return false;
            }
            charCount.put(c, charCount.get(c) - 1);
        }
        return true;
    }

    public boolean isAnagram2(String s1, String s2) {
        if (s1.length() != s2.length()) {
            return false;
        }
        Queue<Character> que1 = new PriorityQueue<>();
        Queue<Character> que2 = new PriorityQueue<>();
        for (int i = 0; i < s1.length(); i++) {
            que1.add(s1.charAt(i));
            que2.add(s2.charAt(i));
        }
        while (que1.size() > 0) {
            if (!que1.poll().equals(que2.poll())) {
                return false;
            }
        }
        return true;
    }

    @org.junit.jupiter.api.Test
    public void doTest() {
        assertEquals(false, isAnagram("cat", "rat"));
        assertEquals(true, isAnagram("anagram", "nagaram"));
    }

}

File: py-algorithm/string/valid_anagram.py
import collections
from collections import Counter

import unittest

"""
    https://leetcode.com/problems/valid-anagram/
    
    Hai chuỗi s1, s2 được gọi là anagram nếu chúng có các ký tự bằng nhau (không tính tứ tự)
    
    Ví dụ: "abc" và "bca" là anagram
"""


class Solution(unittest.TestCase):

    def isAnagram_1(self: str, s2: str) -> bool:
        return Counter(self) == Counter(s2)

    # def isAnagram_3(s1: str, s2: str) -> bool:
    #     mark = 26 * [0]
    #     if len(s1) != len(s2):
    #         return False
    #
    #     for c1, c2 in zip(s1, s2):
    #         mark[ord(c1) - ord('a')] += 1
    #         mark[ord(c2) - ord('a')] -= 1
    #
    #     return all(v == 0 for v in mark)

    def isAnagram_2(self, s1: str, s2: str) -> bool:
        mark = collections.defaultdict(int)
        for k in s1: mark[k] += 1
        for k in s2: mark[k] += -1
        return all(v == 0 for v in mark.values())

    def isAnagram(self, s1: str, s2: str) -> bool:
        mark1 = [0] * 26
        mark2 = [0] * 26
        for c in s1: mark1[c - 'a'] += 1
        for c in s2: mark2[c - 'a'] += 1
        return mark1 == mark2

    def test(self):
        self.assertEqual(False, self.isAnagram("cat", "rat"))
        self.assertEqual(True, self.isAnagram("anagram", "nagaram"))


if __name__ == "__main__":
    unittest.main()
